reco_utils/dataset/movielens.py

```diff
@@ -377,48 +424,60 @@ def _get_schema(header, schema):
     return schema
 
 
-def _load_datafile(size, local_cache_path):
-    """ Download and extract file """
+def _maybe_download_and_extract(size, local_cache_path):
+    """Download and extract MovieLens rating and item datafiles if they don’t already exist.
 
+    Args:
+        size (str): Size of the data to load. One of ("100k", "1m", "10m", "20m").
+        local_cache_path (str): Path where to cache the zip file locally.
+            If None, use default, which is "data/ml-{size}/ml.zip"
+
+    Returns:
+        str: rating datafile path,
+        str: item datafile path.
+    """
     if size not in DATA_FORMAT:
         raise ValueError(ERROR_MOVIE_LENS_SIZE)
-    if not local_cache_path.endswith(".zip"):
+    if local_cache_path is None:
+        default_dir = os.path.join("data", "ml-{}".format(size))
+        local_cache_path = os.path.join(default_dir, "ml.zip")
+    elif not local_cache_path.endswith(".zip"):
         raise ValueError(ERROR_LOCAL_CACHE_PATH)
 
     path, filename = os.path.split(os.path.realpath(local_cache_path))
-
-    # Make sure a temporal zip file get cleaned up no matter what
-    atexit.register(_clean_up, local_cache_path)
-
-    maybe_download(
-        "http://files.grouplens.org/datasets/movielens/ml-" + size + ".zip",
-        filename,
-        work_directory=path,
-    )
+    if not os.path.exists(path):
+        os.makedirs(path)
 
     _, dataname = os.path.split(DATA_FORMAT[size].path)
     datapath = os.path.join(path, dataname)
     _, item_dataname = os.path.split(DATA_FORMAT[size].item_path)
     item_datapath = os.path.join(path, item_dataname)
 
-    with ZipFile(local_cache_path, "r") as z:
-        with z.open(DATA_FORMAT[size].path) as zf, open(datapath, 'wb') as f:
-            shutil.copyfileobj(zf, f)
-        with z.open(DATA_FORMAT[size].item_path) as zf, open(item_datapath, 'wb') as f:
-            shutil.copyfileobj(zf, f)
+    if not os.path.exists(dataname) or not os.path.exists(item_datapath):
+        download_movielens(size, filename, path)
+        _extract(size, local_cache_path, datapath, item_datapath)
+        # Remove downloaded zip file since we don't need anymore.
+        os.remove(local_cache_path)
 
-    _clean_up(local_cache_path)
+    return datapath, item_datapath
 
-    # Make sure a temporal data file get cleaned up when done
-    atexit.register(_clean_up, datapath)
-    atexit.register(_clean_up, item_datapath)
 
-    return datapath, item_datapath
+def download_movielens(size, dest_filename, dest_path="."):
```

Comment:
i don't think we gain anything by encapsulating this as a wrapper function

Reply:
I think we need this function, same in criteo. There are two reasons, one is single responsibility and the second is because this is a function that can be used independently for just download the dataset

Reply:
@miguelgfierro Remember you mentioned someone asked to have this standalone download function. What's the use-case for this compare to him directly downloads the data from MovieLens' url?

Reply:
but we already have a function that provides this responsibility, wrapping it with another function that literally reformats a variable is redundant.

what advantage is this function versus calling maybe_download directly?

Reply:
@gramhagen One good reason of having the download wrapper is, to make sure the users download the same data as the repo's data-loader. 

Please find more context for this from my reply to the earlier comment.

Reply:
> but we already have a function that provides this responsibility, wrapping it with another function that literally reformats a variable is redundant. what advantage is this function versus calling maybe_download directly?

not exactly the same responsibility we are making super easy to download a dataset. I can do `download_movielens("100k")` I don't need to know which url I have to go. 

Also, there was a person asking for this. 

Then finally, before, we had a function called load_dataset, which downloads, extracts and loads the data. That's 3 responsibilities in one function.

Reply:
heh, ok. i'm fine either way honestly. it makes sense to create a seam between pulling data and loading it so we can easily change how we are pulling/extracting/caching and how we load the data for different formats. But in my mind pulling/extracting/caching is a single responsibility for getting the data until it is generalized beyond one dataset. I'm hoping we can offload much of this kind of work to something like https://github.com/intake/intake in the future though, since it's only a necessity to do the more interesting recommendation work

Reply:
Just to chime in on one use case, we need a separate download_movielens function for running aml example where the dataset is downloaded to local cache then uploaded to Azure datastore.

Reply:
@jingyanwangms Usually a user will need to do feature extraction after dataset download (especially when using item or user features), and if he/she runs hyperdrive 100 jobs, the same feature extraction will be done 100 times... Is there any reason for uploading a raw datafile instead of uploading the preprocessed data? If so, isn't it better to run the downloader script from aml compute?

fyi, in wide & deep aml-hyperdrive example, we upload the preprocessed data (pandas dataframe) to the datastore.